cudamapper/src/main.cu

```diff
@@ -397,415 +182,341 @@ void align_overlaps(DefaultDeviceAllocator allocator,
+    std::vector<std::thread> writer_threads;
+    for (int32_t i = 0; i < writer_threads_per_device; ++i)
+    {
+        writer_threads.emplace_back(writer_thread_function,
                                     device_id,
-                                    parameters.all_to_all,
-                                    parameters.filtering_parameter,
-                                    cuda_stream);
-        }
-
-        //Main loop
-        for (const IndexDescriptor& target_range : query_target_range.target_ranges)
-        {
-
-            auto target_start_index = target_range.first_read();
-            auto target_end_index   = target_range.first_read() + target_range.number_of_reads();
-            {
-                CGA_NVTX_RANGE(profiler, "generate_target_index");
-                target_index = get_index(allocator,
-                                         *target_parser,
-                                         target_start_index,
-                                         target_end_index,
-                                         parameters.k,
-                                         parameters.w,
-                                         device_id,
-                                         true,
-                                         parameters.filtering_parameter,
-                                         cuda_stream);
-            }
-            {
-                CGA_NVTX_RANGE(profiler, "generate_matcher");
-                matcher = Matcher::create_matcher(allocator,
-                                                  *query_index,
-                                                  *target_index,
-                                                  cuda_stream);
-            }
-            {
+                                    std::ref(application_parameters),
+                                    std::ref(overlaps_and_cigars_to_write),
+                                    std::ref(output_mutex));
+    }
 
-                OverlapperTriggered overlapper(allocator, cuda_stream);
-                CGA_NVTX_RANGE(profiler, "generate_overlaps");
+    // keep processing batches of indices until there are none left
+    cga_optional_t<BatchOfIndices> batch_of_indices = batches_of_indices.get_next_element();
+    while (batch_of_indices) // if optional is empty that means that there are no more batches to process and the thread can finish
+    {
+        const int64_t batch_number         = number_of_processed_batches.fetch_add(1); // as this is not called atomically with get_next_element() the value does not have to be completely accurate, but this is ok as the value is only use for displaying progress
+        const std::string progress_message = "Device " + std::to_string(device_id) + " took batch " + std::to_string(batch_number + 1) + " out of " + std::to_string(number_of_total_batches) + " batches in total\n";
+        std::cerr << progress_message; // TODO: possible race condition, switch to logging library
```

Comment:
Will this be fixed in a separate PR? I'm wondering if we should remove it for the time being. What do you think?

Reply:
I would leave it as it is for now. In the case of a race condition progress message would still be visible, it would just be interleaved.
I find it very useful to have some form of progress indicator, especially for larger genomes. Initially I was only writing "new batch started" but then realized having some number is really convenient, even if it is not completely accurate

Reply:
yes, makes sense.